javacutil/src/main/java/org/checkerframework/javacutil/ElementUtils.java

```diff
@@ -239,6 +264,43 @@ public static boolean isCompileTimeConstant(Element elt) {
                 && ((VariableElement) elt).getConstantValue() != null;
     }
 
+    /**
+     * Checks whether a given element came from a source file.
+     *
+     * <p>By contrast, {@link ElementUtils#isElementFromByteCode(Element)} returns true if there is
+     * a classfile for the given element, even if there is also a source file.
+     *
+     * @param element the element to check, or null
+     * @return true if a source file containing the element is being compiled
+     */
+    public static boolean isElementFromSourceCode(@Nullable Element element) {
+        if (element == null) {
+            return false;
+        }
+        if (element instanceof ClassSymbol) {
+            return isElementFromSourceCodeImpl((ClassSymbol) element);
+        }
+        return isElementFromSourceCode(element.getEnclosingElement());
+    }
+
+    /**
+     * Checks whether a given ClassSymbol came from a source file.
+     *
+     * <p>By contrast, {@link ElementUtils#isElementFromByteCode(Element)} returns true if there is
+     * a classfile for the given element, even if there is also a source file.
+     *
+     * @param symbol the class to check
+     * @return true if a source file containing the class is being compiled
+     */
+    private static boolean isElementFromSourceCodeImpl(ClassSymbol symbol) {
+        // This is a bit of a hack to avoid treating JDK as source files. JDK files' toUri() method
```

Comment:
I wish I knew. I'd speculate it *might* have something to do with the annotated JDK, but no evidence for that. It would probably be interesting to look into.